## Итоговый тест
### Вариант - 5

## Задача 1

### Исходные данные:
|          | A    | B    | C    | D    |
|----------|------|------|------|------|
| Количество | 300  | 250  | 50   | 200  |
| Текущая цена | 50   | 75   | 25   | 100  |
| Цена на конец года | 70   | 90   | 20   | 120  |
 
Корректировка цены: `50 * N = 50 * 5 = 250`

---

### Шаг 1. Корректировка цен
Добавляем 250 к каждой цене:

**Текущие цены:**
- A: `50 + 250 = 300`
- B: `75 + 250 = 325`
- C: `25 + 250 = 275`
- D: `100 + 250 = 350`

**Цены на конец года:**
- A: `70 + 250 = 320`
- B: `90 + 250 = 340`
- C: `20 + 250 = 270`
- D: `120 + 250 = 370`

---

### Шаг 2. Расчет доходности каждой облигации
Формула:  
`Доходность = ((Цена_конец_года - Текущая_цена) / Текущая_цена) * 100%`

- **A**:  
  `((320 - 300) / 300) * 100% = (20 / 300) * 100% ≈ 6.67%`
- **B**:  
  `((340 - 325) / 325) * 100% = (15 / 325) * 100% ≈ 4.62%`
- **C**:  
  `((270 - 275) / 275) * 100% = (-5 / 275) * 100% ≈ -1.82%`
- **D**:  
  `((370 - 350) / 350) * 100% = (20 / 350) * 100% ≈ 5.71%`

---

### Шаг 3. Расчет общей стоимости портфеля
**Текущая стоимость:**  
`300*300 + 250*325 + 50*275 + 200*350 = 255,000`

**Стоимость на конец года:**  
`300*320 + 250*340 + 50*270 + 200*370 = 268,500`

---

### Шаг 4. Расчет доходности портфеля
`Доходность = ((268,500 - 255,000) / 255,000) * 100% ≈ 5.29%`

---

### Итоговый ответ:
Доходность портфеля: **5.29%**

<span style="color:yellow">---------------------------------------------------------------------------------------------------------------------</span>

### Задача 2

#### Исходные данные:
| R_A    | -4    | 3     | 10    | 17    | 24    |
|--------|-------|-------|-------|-------|-------|
| R_B    | 24    | 16,5  | 9     | 1,5   | -6    |
| Вероятность | 0,0625 | 0,25  | 0,375 | 0,25  | 0,0625 |

Номер варианта: **5**  
Корректировка: `0.1 * N = 0.1 * 5 = 0.5`

---

#### Шаг 1. Корректировка доходностей
Добавляем 0.5 к каждому значению доходности:

**Скорректированные доходности:**
- **Актив A (R_A'):**
  - -4 + 0.5 = -3.5
  - 3 + 0.5 = 3.5
  - 10 + 0.5 = 10.5
  - 17 + 0.5 = 17.5
  - 24 + 0.5 = 24.5

- **Актив B (R_B'):**
  - 24 + 0.5 = 24.5
  - 16.5 + 0.5 = 17.0
  - 9 + 0.5 = 9.5
  - 1.5 + 0.5 = 2.0
  - -6 + 0.5 = -5.5

---

#### Шаг 2. Расчет математических ожиданий
Формула:  
`E(R) = Σ (R_i * P_i)`

- **E(R_A'):**
  `(-3.5)*0.0625 + 3.5*0.25 + 10.5*0.375 + 17.5*0.25 + 24.5*0.0625 = 10.5`

- **E(R_B'):**
  `24.5*0.0625 + 17.0*0.25 + 9.5*0.375 + 2.0*0.25 + (-5.5)*0.0625 = 9.5`

---

#### Шаг 3. Расчет ковариации
Формула:  
`Cov(A,B) = Σ P_i * (R_Ai' - E(R_A')) * (R_Bi' - E(R_B'))`

Вычисления:
1. (-3.5 - 10.5)*(24.5 - 9.5)* 0.0625 = (-14)*15*0.0625 = -13.125
2. (3.5 - 10.5)*(17.0 - 9.5)* 0.25 = (-7)*7.5*0.25 = -13.125
3. (10.5 - 10.5)*(9.5 - 9.5)* 0.375 = 0*0*0.375 = 0
4. (17.5 - 10.5)*(2.0 - 9.5)* 0.25 = 7*(-7.5)*0.25 = -13.125
5. (24.5 - 10.5)*(-5.5 - 9.5)* 0.0625 = 14*(-15)*0.0625 = -13.125

`Cov(A,B) = -13.125 + (-13.125) + 0 + (-13.125) + (-13.125) = -52.5`

---

#### Шаг 4. Расчет стандартных отклонений
Формула:  
`σ(R) = √(Σ P_i * (R_i' - E(R'))^2)`

- **σ(R_A'):**
  `√[0.0625*(-14)^2 + 0.25*(-7)^2 + 0.375*0^2 + 0.25*7^2 + 0.0625*14^2] = √[12.25 + 12.25 + 0 + 12.25 + 12.25] = √49 = 7`

- **σ(R_B'):**
  `√[0.0625*15^2 + 0.25*7.5^2 + 0.375*0^2 + 0.25*(-7.5)^2 + 0.0625*(-15)^2] = √[14.0625 + 14.0625 + 0 + 14.0625 + 14.0625] = √56.25 = 7.5`

---

#### Шаг 5. Расчет коэффициента корреляции
Формула:  
`ρ(A,B) = Cov(A,B) / (σ(R_A') * σ(R_B'))`

`ρ(A,B) = -52.5 / (7 * 7.5) = -52.5 / 52.5 = -1`

---

### Итоговые результаты:
1. Ковариация активов A и B: **-52.5**  

2. Коэффициент корреляции: **-1**  

<span style="color:yellow">---------------------------------------------------------------------------------------------------------------------</span>

## Задача 3

**Доли в портфеле:**
- x₁ = 5/16 = 0.3125
- x₂ = 3/8 = 0.375
- x₃ = 5/16 = 0.3125

**Доходности в соответствии с вариантом:**
- μ₁ = 16% + 0.5% = 16.5%
- μ₂ = 24% + 0.5% = 24.5%
- μ₃ = -8% + 0.5% = -7.5%


In [8]:
x = [0.3125, 0.375, 0.3125] # вектор долей ценных бумаг в портфеле

**Матрица ковариаций:**

In [ ]:
V = [[5.01, 0.4, 1.47],
[0.4, 9.8, 0.7],
[1.47, 0.7, 5.29]]


#### 1. Ожидаемая доходность портфеля

μₚ = x₁μ₁ + x₂μ₂ + x₃μ₃

μₚ = 0.3125×16.5% + 0.375×24.5% + 0.3125×(-7.5%)  
   = 5.15625% + 9.1875% - 2.34375%  
   = **12.0%**

#### 2. Риск портфеля (среднеквадратичное отклонение)
σₚ = √(xᵀVx)

##### Шаг 1. Умножение матрицы на вектор
Вычисляем Vx:

In [3]:
import numpy as np

In [4]:
x = [0.3125, 0.375, 0.3125]

In [7]:
Vx = np.dot(V, x)
Vx

array([2.175  , 4.01875, 2.375  ])


### Шаг 2. Умножение векторов
xᵀVx = [0.3125, 0.375, 0.3125] × [2.175, 4.01875, 2.375]  
      = 0.3125×2.175 + 0.375×4.01875 + 0.3125×2.375  
      = 0.6797 + 1.5070 + 0.7422  
      = 2.9289

### Шаг 3. Найдем среднеквадратичное отклонение
σₚ = √2.9289 ≈ **1.7114%**

## Вывод
| Показатель            | Значение |
|-----------------------|----------|
| Ожидаемая доходность  | 12.0%    |
| Риск портфеля (σ)     | 1.7114%  |


<span style="color:yellow">---------------------------------------------------------------------------------------------------------------------</span>

## Задача 4
### Исходные данные

**Корректировка доходности:** 0.1 × 5 = 0.5%

**Ценные бумаги:**
1. Безрисковая ЦБ:
   - Доходность: r_f = 3% + 0.5% = **3.5%**
   - Риск: σ_f = 0%

2. Рискованная ЦБ A:
   - Доходность: μ_A = 7% + 0.5% = **7.5%**
   - Риск: σ_A = 4%
   - Корреляция с ЦБ B: ρ_AB = 0

3. Рискованная ЦБ B:
   - Доходность: μ_B = 10% + 0.5% = **10.5%**
   - Риск: σ_B = 5%
   - Корреляция с ЦБ A: ρ_AB = 0

**Требуемая доходность портфеля:** μ_p = 7.4% + 0.5% = **7.9%**

### Решение

#### 1. Оптимальные веса рискованных активов

Для активов с нулевой корреляцией веса рассчитываются как:

w_A = [σ_B² × (μ_A - r_f)] / [σ_A² × (μ_B - r_f) + σ_B² × (μ_A - r_f)]
w_B = 1 - w_A

Подставляем значения:

w_A = [5² × (7.5 - 3.5)] / [4² × (10.5 - 3.5) + 5² × (7.5 - 3.5)]
    = (25 × 4) / (16 × 7 + 25 × 4) 
    = 100 / (112 + 100) 
    = 100/212 ≈ 0.4717 (47.17%)

w_B = 1 - 0.4717 = 0.5283 (52.83%)


#### 2. Характеристики рискованного портфеля

Доходность рискованной части:

μ_risky = w_A × μ_A + w_B × μ_B 
        = 0.4717 × 7.5% + 0.5283 × 10.5% ≈ 9.08%


Риск рискованной части:

σ_risky = √[w_A² × σ_A² + w_B² × σ_B² + 2 × w_A × w_B × ρ_AB × σ_A × σ_B]
        = √[0.4717² × 16 + 0.5283² × 25 + 0] 
        ≈ √[3.56 + 6.98] ≈ √10.54 ≈ 3.25%


#### 3. Доля рискованных активов (y)

Из уравнения капитальной линии:

μ_p = r_f + y × (μ_risky - r_f)
7.9 = 3.5 + y × (9.08 - 3.5)
y = (7.9 - 3.5)/(9.08 - 3.5) ≈ 4.4/5.58 ≈ 0.7885 (78.85%)


#### 4. Структура портфеля

| Компонент       | Доля      | Расчет                     |
|----------------|-----------|----------------------------|
| Безрисковый    | 21.15%    | 1 - 0.7885 = 0.2115        |
| ЦБ A           | 37.19%    | 0.7885 × 47.17% ≈ 0.3719   |
| ЦБ B           | 41.66%    | 0.7885 × 52.83% ≈ 0.4166   |

#### 5. Риск портфеля

σ_p = y × σ_risky = 0.7885 × 3.25% ≈ 2.56%


### Итоговые характеристики
- **Ожидаемая доходность:** 7.9%
- **Риск (σ):** 2.56%
- **Распределение:**
  - Безрисковые: 21.15%
  - ЦБ A: 37.19%
  - ЦБ B: 41.66%

### 